### PR TITLE
Made changes needed for deploying secure Che with OAuth provider configured on Minishift

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
There are several commits that improve deploying secure Che on Minishift or any other OpenShift cluster that is not local OCP:
* Add settings of `-e` mode in deploy_che.sh script
It is needed because previously any fatal error doesn't break the deploying process. As result, there is a message that Che is Deployed but actually, deployment may be is in inconsistency state.
* Add an ability to set up oauth provided from deploy_che.sh with the argument
There are a lot of missing default values for environment variable. This commit adds needed env vars with defaults and add respecting of `--setup-ocp-oauth` argument
* Add more info about current deploying state
Just few more message about current deploying state to make it clear of a user.
* Add an ability to specify self-signed certificate during deploying che
Previously certificate should be located in file `/var/lib/origin/openshift.local.config/master/ca.crt` and it works only for OCP on docker. Now there is an ability to specify certificate content with `OPENSHIFT_CERT` env var or path to cert file with `OPENSHIFT_CERT_PATH` env var
* Remove supressing output/input during execution
In case of deploying on minishift with self-signed certificate, `oc login ... > /dev/null` command hangs because of waiting approval that oc may trust self-signed certificate. So, `> /dev/null` is removed and now user is able to type `y` to allow trusting self-signed certificate
* Create or update oauth client instead of try to create a new one
Previously the only first deployment works because `oauth-client` is created at cluster scope, and the next tries fails with error `oauth client already exists`. Using of `oc apply` allows create or update if it exists.

I was not able to test it on minishift because of the following bug https://github.com/openshift/origin/issues/20716. I tested it by deploying Che with `deploy_che.sh` on Local OCP:
```bash
export OPENSHIFT_CERT=$(cat /var/lib/origin/openshift.local.config/master/ca.crt)
./deploy_che.sh --multiuser --no-pull --rolling --secure --setup-ocp-oauth
```

### What issues does this PR fix or reference?
It is related(but not needed) to https://github.com/eclipse/che/issues/10910.

#### Release Notes
N/A

#### Docs PR
N/A

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by l0rd** [comment]
@sleshchenko wouldn't it be better to work on the [minishift addon](https://github.com/minishift/minishift/tree/master/addons/che) instead? The addon should be the preferred/supported way to deploy Che on minishift but currently it doesn't support multi-user https://github.com/minishift/minishift/issues/2746. Working on that would have a lot of value and would help us to accelerate the decomission of single user Che.

**Comment by sleshchenko** [comment]
@l0rd I would glad to but

> currently it doesn't support multi-user minishift/minishift#2746.

it's a key phrase for me. Scripts give a flexible of mixing different configuration multiuser/secure/oauth_provider/che-plugin-registry. I'm not aware of how to implement the same with minishift addon. So, I would merge this PR if changes are OK since these scripts are used internally and I didn't spend much time for that, I just volunteered some time for these changes. Otherwise, we need a register issue for making it possible to do the same on minishift addon.
P.S. Note that `deploy-che.sh` may be used for deploying Che on any cluster (it should not be OCP or minishift). Minishift addon doesn't allow deploying on any OpenShift cluster. Because of that I think these scripts are valuable until Che Operator is not ready.

**Comment by l0rd** [comment]
@sleshchenko in the title you refer to minishift, in the subject you talk about minikube. In your last comment you talk about any OpenShift cluster. I am a little bit lost :smiley:  

Anyway I wanted to raise the fact that working on supporting multi-user Che on minishift is one of the priorities of this train. Maybe the next thing you could work on?

**Comment by eivantsov** [comment]
@l0rd supporting multiuser Che on Minishift can be anything from working on MiniShift addon (+ all extras like https support and OpenShift oAuth) to releasing an Operator that has it all (it doe snot matter where it runs, as long as it's a healthy k8s/openshift cluster).

**Comment by sleshchenko** [comment]
@l0rd 

> @sleshchenko in the title you refer to minishift, in the subject you talk about minikube. In your last comment you talk about any OpenShift cluster. I am a little bit lost smiley

Good catch ) Fixed. Get rid of `minikube`

> Anyway I wanted to raise the fact that working on supporting multi-user Che on minishift is one of the priorities of this train. Maybe the next thing you could work on?

I think it's question to my team lead but not to me

**Comment by l0rd** [comment]
@eivantsov on minishift we should support the addon only. The addon is using the current upstream templates but can and should use the operator in the future. Can we agree on that?

@sleshchenko yes please discuss that with @skabashnyuk

**Comment by riuvshin** [comment]
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1289/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1289//Selenium_20tests_20report/)
docker image: eclipseche/che-server:12166
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by dmytro-ndp** [comment]
[Previous E2E tests build](https://github.com/eclipse/che/pull/12166#issuecomment-446331107) was aborted because it timed out (after 300 minutes) beside it was going not bad:
```
Passed: 639, failed: 17, skipped: 11.
```

**Comment by dmytro-ndp** [comment]
ci-test

**Comment by riuvshin** [comment]
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1290/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1290//Selenium_20tests_20report/)
docker image: eclipseche/che-server:12166
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by dmytro-ndp** [comment]
ci-test

**Comment by riuvshin** [comment]
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1291/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1291//Selenium_20tests_20report/)
docker image: eclipseche/che-server:12166
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by sleshchenko** [comment]
@dmytro-ndp I assume test report is good enough. Please notify me whether I can merge it or you want to test it locally with https enabled.

**Comment by dmytro-ndp** [comment]
@sleshchenko: I would test it locally before you merge it into master.

**Comment by rhopp** [comment]
@sleshchenko According to the wrong certificate in minishift... I found out, that minishift devs already tackled this issue ([originally reported here](https://github.com/openshift/origin/issues/19844)) and they created [minishift addon](https://github.com/minishift/minishift/blob/master/addons/registry-route/registry-route.addon) fixing this. Apparently simmilar steps should be fine to apply even to cluster started by `oc cluster up`

**Comment by sleshchenko** [comment]
@rhopp Thanks for sharing, will take a look. 
@dmytro-ndp You may use it to test my changes locally.

**Comment by eivantsov** [comment]
That solution does not work for us. The only workaround is regenerating cert and updates router with it

**Comment by dmytro-ndp** [comment]
@sleshchenko: good job!
I have successfully deployed Eclipse Che in https mode to OS Origin 3.9 (from minishift) and run workspace:
screenshot from 2018-12-12 17-01-11

**Review by dmytro-ndp — APPROVED** [review]

**Pull request merged**
